admin_guide/preventing_node_placement.adoc

```diff
@@ -0,0 +1,74 @@
+= Preventing Specific Node Placement
+{product-author}
+{product-version}
+:data-uri:
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+As a cluster administrator, you can prevent users from targeting specific nodes
+when scheduling pods.  
+
+== Constraining Pod Placement
+
+To prevent users from targeting specific nodes, modify the OpenShift master
+configuration, and the Kubernetes resource configuration.
+
+Adding a label to the `nodeSelectorLabelBlacklist` section of a master
+configuration file prevents pods being deployed onto the nodes with that label.
```

Comment:
This isn't correct - adding a label to the blacklist prevents Pods from being created if their NodeSelector carries any of the blacklisted labels. If the NodeSelector is empty or carries no blacklisted labels, then the scheduler may still place the Pod on a Node that has a blacklisted label.

Say you have a Node with these labels:
```
  - kubernetes.io/hostname: node1
  - foo: bar1
  - baz: quux
```

Also, your `NodeSelectorBlacklist` looks like this:
```
  - kubernetes.io
  - foo
  - shmekky
```

Users which lack the `pods/binding` permission will be able to:

* Create Pods with empty NodeSelector fields
* Create Pods with a NodeSelector label which isn't blacklisted
  * This can be a label that exists on a Node, like `baz: quux`
  * One could also specify a label that doesn't exist on a Node, and it should still work, such as `fizz: buzz` in this example

Users which lack the `pods/binding` permission will **not** be able to:

* Create Pods which specify a NodeName
  * The NodeName field is pretty much the same as a NodeSelector of `kubernetes.io/hostname: node1` in this example
* Create Pods with a NodeSelector label that **is** blacklisted
  * This can be a label that exists on a Node, like `foo: bar`
  * Labels which don't exist on any Nodes but are in the blacklist will also be prohibited, such as `shmekky: blekky`

For Users, ServiceAccounts, or Groups which *do* have the `pods/binding` permission, the NodeName constraint and NodeSelector label blacklist will have no effect.

In all of these cases, the scheduler may *still* correctly schedule a Pod on a Node which bears a blacklisted label as long as it's in the set of schedulable Nodes. That set is determined by a number of other criteria, such as project-level NodeSelectors, etc.

One more example that might make the function of this plugin more clear. Say your cluster is spread across five datacenter "regions": "us-east", "us-central", "us-west", and across the ocean "apac-east" and "apac-west". Nodes which are physically located in each region are labelled accordingly, i.e. `region: us-west` for example. In addition, the specific regions are grouped by labels into "super-regions": `superregion: us` and `superregion: apac`

As the cluster administrator, you may want to group Users by region and configure their projects so that their new Pods are scheduled on Nodes which are geographically local to the User. For Users in the US, their projects are created with the project-level NodeSelector `superregion: us`, and for users in the Asia-Pacific region, their projects are created with project-level NodeSelector `superregion: apac`.

For resource management reasons, you want to make sure load is spread evenly across all Nodes in a super-region. You also want to allow administrators to place new Pods in particular regions for monitoring purposes. To prevent Users from weighting a particular region, you add the `region` label to the `NodeSelectorLabelBlacklist`. Now whenever a User in the US - who by policy doesn't have `pods/binding` permission - creates a Pod, it gets scheduled to a Node in any of the three regions that overlap with the `us` super-region with equal likelihood, according to the best-fit criteria for your cluster. If the User tries to target a `region` with their Pod, it will fail. If they try again, but omit the NodeSelector on their Pod, it may be scheduled on the region they tried to target, because the `us` super-region is set as the project-level NodeSelector, and those Nodes are labeled as part of the `us` super-region.

Your administrators are still able to set the `region` label in their Pod NodeSelector fields because the `cluster-admin` role has the `pods/binding` permission.

I hope that makes more sense.